src/validation/func.rs

```diff
@@ -0,0 +1,773 @@
+
+use common::stack::StackWithLimit;
+use common::{BlockFrame, BlockFrameType};
+
+/// Constant from wabt' validator.cc to skip alignment validation (not a part of spec).
+const NATURAL_ALIGNMENT: u32 = 0xFFFFFFFF;
+
+/// Maximum number of entries in value stack.
+const DEFAULT_VALUE_STACK_LIMIT: usize = 16384;
+/// Maximum number of entries in frame stack.
+const DEFAULT_FRAME_STACK_LIMIT: usize = 1024;
+
+/// Function validation context.
+struct FunctionValidationContext<'a> {
+	/// Wasm module
+	module: &'a ModuleContext,
+	/// Current instruction position.
+	position: usize,
+	/// Local variables.
+	locals: &'a [ValueType],
+	/// Value stack.
+	value_stack: StackWithLimit<StackValueType>,
+	/// Frame stack.
+	frame_stack: StackWithLimit<BlockFrame>,
+	/// Function return type. None if validating expression.
+	return_type: Option<BlockType>,
+	/// Labels positions.
+	labels: HashMap<usize, usize>,
+}
+
+/// Value type on the stack.
+#[derive(Debug, Clone, Copy)]
+enum StackValueType {
+	/// Any value type.
+	Any,
+	/// Any number of any values of any type.
+	AnyUnlimited,
+	/// Concrete value type.
+	Specific(ValueType),
+}
+
+/// Function validator.
+pub struct Validator;
+
+/// Instruction outcome.
+#[derive(Debug, Clone)]
+enum InstructionOutcome {
+	/// Continue with next instruction.
+	ValidateNextInstruction,
+	/// Unreachable instruction reached.
+	Unreachable,
+}
+
+impl Validator {
+	pub fn validate_function(
+		module: &ModuleContext,
+		func: &Func,
+		body: &FuncBody,
+	) -> Result<(), Error> {
+		let (params, result_ty) = module.require_function_type(func.type_ref())?;
+
+		// locals = (params + vars)
+		let mut locals = params.to_vec();
+		locals.extend(
+			body.locals()
+				.iter()
+				.flat_map(|l| repeat(l.value_type())
+				.take(l.count() as usize)
+			),
+		);
+
+		let mut context = FunctionValidationContext::new(
+			&module,
+			&locals,
+			DEFAULT_VALUE_STACK_LIMIT,
+			DEFAULT_FRAME_STACK_LIMIT,
+			result_ty,
+		);
+
+		context.push_label(BlockFrameType::Function, result_ty)?;
+		Validator::validate_function_block(&mut context, body.code().elements())?;
+		while !context.frame_stack.is_empty() {
+			context.pop_label()?;
+		}
+
+		Ok(())
+	}
+
+	fn validate_function_block(context: &mut FunctionValidationContext, body: &[Opcode]) -> Result<(), Error> {
+		let body_len = body.len();
+		if body_len == 0 {
+			return Err(Error("Non-empty function body expected".into()));
+		}
+
+		loop {
+			let opcode = &body[context.position];
+			match Validator::validate_instruction(context, opcode)? {
+				InstructionOutcome::ValidateNextInstruction => (),
+				InstructionOutcome::Unreachable => context.unreachable()?,
+			}
+
+			context.position += 1;
+			if context.position >= body_len {
+				return Ok(());
+			}
+		}
+	}
+
+	fn validate_instruction(context: &mut FunctionValidationContext, opcode: &Opcode) -> Result<InstructionOutcome, Error> {
+		debug!(target: "validator", "validating {:?}", opcode);
+		match opcode {
```

Comment:
if this already in the diff and pr is about validation won't you mind changing to `match *opcode { ... }`

Reply:
yeah sure